File: packages/Recognizers/ConversationLanguageUnderstanding/dotnet/CluExtensions.cs
using System.Collections.Generic;
using System.Globalization;
using Microsoft.Bot.Builder;
using Newtonsoft.Json.Linq;

namespace Microsoft.Bot.Components.Recognizers.CLURecognizer
{
    internal static class CluExtensions
    {
        internal static IDictionary<string, IntentScore> ExtractIntents(this JObject cluResult)
        {
            var result = new Dictionary<string, IntentScore>();

            if (cluResult?["intents"] != null && cluResult["intents"] is JArray cluIntents)
            {
                foreach (var intent in cluIntents)
                {
                    var category = intent["category"] !.ToString();
                    var confidenceScore = intent["confidenceScore"] !.ToString();

                    result.Add(
                        NormalizedValue(category),
                        new IntentScore
                        {
                            Score = confidenceScore == null ? 0.0 : double.Parse(confidenceScore, CultureInfo.InvariantCulture),
                        });
                }
            }

            return result;
        }

        internal static JObject ExtractEntities(this JObject cluResult)
        {
            var result = new JObject();

            if (cluResult?["entities"] != null && cluResult["entities"] is JArray cluEntities)
            {
                foreach (var entity in cluEntities)
                {
                    var category = entity["category"] !.ToString();

                    if (result[NormalizedValue(category)] == null)
                    {
                        result[NormalizedValue(category)] = new JArray(entity);
                        continue;
                    }

                    ((JArray)result[NormalizedValue(category)] !).Add(entity);
                }
            }

            return result;
        }

        private static string NormalizedValue(string value) => value.Replace('.', '_').Replace(' ', '_');
    }
}

File: packages/Recognizers/ConversationLanguageUnderstanding/js/src/cluExtensions.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/* eslint-disable  @typescript-eslint/no-explicit-any */

import { IntentScore } from 'botbuilder';

/**
 * Utility class for CLU Results.
 */
export class CluExtensions {
  /**
   * Extract intents from a CLU Result.
   * @param cluResult The CLU Result.
   * @returns An object with the extracted intents.
   */
  static extractIntents(
    cluResult: Record<string, any>
  ): Record<string, IntentScore> {
    const result: Record<string, IntentScore> = {};
    if (!!cluResult?.intents && Array.isArray(cluResult.intents)) {
      for (const intent of cluResult.intents) {
        result[this.normalizedValue(intent.category)] = {
          score: !intent.confidenceScore
            ? 0.0
            : Number.parseFloat(intent.confidenceScore),
        };
      }
    }
    return result;
  }

  /**
   * Extract entities from a CLU Result.
   * @param cluResult The CLU Result.
   * @returns An object with the extracted entities.
   */
  static extractEntities(cluResult: Record<string, any>): Record<string, any> {
    const result: Record<string, any> = {};
    if (!!cluResult?.entities && Array.isArray(cluResult.entities)) {
      for (const entity of cluResult.entities) {
        const normalizedCategory = this.normalizedValue(entity.category);
        if (!result[normalizedCategory]) {
          result[normalizedCategory] = [];
        }

        result[normalizedCategory].push(entity);
      }
    }
    return result;
  }

  private static normalizedValue(value: string) {
    return value.replace('.', '_').replace(' ', '_');
  }
}
